XML parsing for LegacyTrackerTopology

pieterdavid opened the pull request:
See code for details.

I started with a standalone C++ executable, the parsing works so I think it's technically ok, but there may be typos/bugs/... so please have a look ;)

pieterdavid committed:
Parse trackerParameters.xml for LegacyTrackerTopology

The name of the xml file to use is an argument to
LegacyTrackerTopology::getTrackerTopology (with no default value)
Parsing is done using Xerces-C++

pieterdavid committed:
Pass trackerParameters.xml path to getTrackerTopology

pieterdavid commented:
The API probably still needs to be changed, making this an `explicit TrackerTopology(const std::string&)` constructor, or a `static TrackerTopology TrackerTopology::fromTrackerParametersXML(const std::string&)` method, or a `namespace StandaloneTrackerTopology{ TrackerTopology fromTrackerParametersXML(const std::string&) }`... Preferences?

OlivierBondu commented on CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C:
```diff
@@ -24,7 +24,7 @@
 
 
-std::unique_ptr<TrackerTopology> tTopo = LegacyTrackerTopology::getTrackerTopology();
+std::unique_ptr<TrackerTopology> tTopo = LegacyTrackerTopology::getTrackerTopology(std::string(gSystem->Getenv("CMSSW_RELEASE_BASE"))+"src/Geometry/TrackerCommonData/data/trackerParameters.xml");
```
Does this means the package` Geometry/TrackerCommonData` needs to be checked out, is it enforced anywhere ?

pieterdavid replied:
No, $CMSSW_RELEASE_BASE is the release path, which is a complete checkout, so it only requires `cmsenv` (the usual one is $CMSSW_BASE).

OlivierBondu commented on DataFormats/TrackerCommon/interface/LegacyTrackerTopology.h:
```diff
@@ -2,96 +2,16 @@
+   * Construct a TrackerTopology from a trackerParameters.xml file
+   * WARNING: this header has been introduced to call a TrackerTopology
+   * object whenever it is not possible to access it through an EventSetup.
+   * Do not use it if that is the case.
```
Maybe we could be more explicit on 'that is the case', re-reading it it's not crystal clear...

pieterdavid replied:
Good point, I rephrased it a bit

OlivierBondu commented on DataFormats/TrackerCommon/src/LegacyTrackerTopology.cc:
```diff
@@ -0,0 +1,233 @@
+    return str;
+  }
+
+  // RAII release the XMLCh*
```
I have no idea what this comment mean :) What's a `XMLCh` ? a XML `char*` ?

pieterdavid replied:
Some Xerces-C++ `char*`-equivalent (Unicode-aware, I think).
RAII is the technique that is also used in smart pointer classes like `std::unique_ptr`, lock handles etc.: make a helper class that acquires/releases a resource in the constructor and destructor, and make an instance as a local variable (the destructor will call the release when this variable goes out of scope, even in case an exception is thrown).
It prevents leaks, and it also avoids having to type a list of declarations at the top, and the corresponding list of cleanup statements at the bottom.

OlivierBondu replied:
oh, ok, nice :)

OlivierBondu commented on DataFormats/TrackerCommon/src/LegacyTrackerTopology.cc:
```diff
@@ -0,0 +1,233 @@
+    auto_XMLString(XMLCh* xch) : m_xch(xch) {}
+    XMLCh* get() const { return m_xch; }
+    ~auto_XMLString() { if ( m_xch ) { XMLString::release(&m_xch); } }
+    // avoid double release: make this class move-only (rule of 5)
```
TIL what rule of 5 means :)

pieterdavid replied:
It used to be called the rule of three until C++11... "whenever any of the destructor, copy constructor, copy assignment operator, move constructor, and move assignment constructor is defined (i.e. non-default), they should all be defined", but at second thought it doesn't belong in a comment, so I took it out again.

OlivierBondu reviewed (approved):
Looks good, only minor comments. I do still have a couple of questions though:
- what about the xml itself, does it already exists ?
- wasn't there also a comment about the `LegacyTrackerTopology` naming being ambiguous with the new pixel ?

I have no preference for the API

pieterdavid committed:
Rename, return by value

now the signature is namespace StandaloneTrackerTopology {
  TrackerTopology fromTrackerParametersXML(const std::string&); }.
Also adapted uses

pieterdavid committed:
update doc and warning

pieterdavid committed:
Simplify comment

pieterdavid commented:
I found several XML files in `Geometry/TrackerCommonData/data`, the most relevant probably `trackerParameters.xml` (Phase0), `PhaseI/trackerParameters.xml` and `PhaseII/trackerParameters.xml`.
The numbers that @alesaggio put there before are identical to those in the Phase0 file.
I changed the namespace to `StandaloneTrackerTopology`, and the method name to `fromTrackerParametersXML` (also made it return by value, that'sa little bit simpler)

OlivierBondu commented:
ok thanks, sound good, all fine on my end!

pieterdavid commented:
Thanks for checking!
I will merge it here then, and ask @alesaggio to update her PR with it

alesaggio commented:
Looks fine by me! Just a question: what about CalibTracker/SiStripChannelGain/test/UsefullCode/TimeHistory/PlotMacro.C? Are we leaving it as it is?

pieterdavid commented:
I also changed that one, don't you see it? It's the first if I look at "files changed"

alesaggio commented:
My bad :) Fine by me then!
